### ISSUE TITLE
has_many :through association does not obey autosave when parent object is new

### ISSUE THREAD

**Issue by scottmatthewman** [rails/rails#19782]
When a model has a `has_many :through` collection set to `autosave`, a new model does not save any members of that association which have been modified. An existing parent model which is saved does, however, make the necessary autosaves.

A self-contained test (tested against ActiveRecord 4.2.1) to illustrate: https://gist.github.com/scottmatthewman/3b01023e58ae908510e8 

Excerpt below:

```ruby
class BugTest < Minitest::Test
  # This test passes
  def test_autosave_has_many_through_on_existing_parent
    catalog = Catalog.create!(title: 'Original Title')
    catalog.title = 'Amended Title'
 
    post = Post.create!
    post.catalogs = [catalog]
    post.save!
 
    assert_equal 1, post.catalogs.count
    assert_equal 'Amended Title', catalog.reload.title
  end
 
  # This test should also pass, but is instead failing
  def test_autosave_has_many_through_on_new_parent
    catalog = Catalog.create!(title: 'Original Title')
    catalog.title = 'Amended Title'
 
    post = Post.new
    post.catalogs = [catalog]
    post.save!
 
    assert_equal 1, post.catalogs.count
    assert_equal 'Amended Title', catalog.reload.title
  end
end
```

**Comment by jfabre** [comment]
Just got the same issue. Is there some kind of workaround other than saving the parent first?

**Comment by scottmatthewman** [comment]
In our codebase, we came up with a workaround using `before_save` that suited our purposes (can't speak for whether it matches others' needs). Example code below (with objects renamed to match the above example objects).

```ruby
class Post
  before save :save_associated_catalogs, on: :create

  private

  def save_associated_catalogs
    catalogs.select(&:changed?).each(&:save)
  end
end
```

**Comment by jfabre** [comment]
Makes sense, thanks!

**Comment by scottmatthewman** [comment]
Closed in error, sorry

**Comment by sgrif** [comment]
I'm assuming the `belongs_to` in `Catalog` was supposed to be a `has_many`

**Comment by scottmatthewman** [comment]
Indeed - thanks for catching that! Have corrected the gist. As you'd expect, that doesn't affect the failing case, though.

**Comment by sgrif** [comment]
No, I'm looking at the failing behavior now. It appears intentional, but I'm seeing if the fix causes any tests to fail.

**Comment by sgrif** [comment]
This actually constitutes a subtle change in behavior (see the commit message for a more thorough explanation), so while this will be fixed in 5.0, I will not be backporting this fix to 4.x

**Comment by scottmatthewman** [comment]
Makes sense. Thanks.
